### PR TITLE
Heightmap additions

### PR DESCRIPTION
**Pull request by BryceBubbles**
I'm working on overhauling the heightmap setup to use the viewport + shader method I mentioned. But that needs quite a bit of thought, and possibly some discussion. I'll get a working example up on my fork some time for you to look at and make suggests.

**In the mean time,** here's the `Apply Image` and `Output Heightmap` nodes, along with some tweaks that were needed to make those work well. 

I added a `Texture` data type, so you may want to check that I did everything needed for that.

I also added another example that uses all the heightmaps nodes so far, but it doesn't use any noise blending, so we may want to keep both examples.

### CODE REVIEW

**Comment by HungryProton** [comment]
Awesome! Looks all good to me, thanks for the PR!

I'd like to add a new method in the Heightmap class along with `get_point(int, int)`, something like `get_point_interpolated(float, float)` to get a smoother result when using a mesh with a higher subdivision level than the texture size : 

image
image

image)

The last picture uses your example with the shader mesh but it's not exactly the same artifact for some reason. I guess the shader already does some kind of interpolation?

I was waiting for your PR before starting working on it but I'd like your opinion on how to proceed (or maybe you already have the code for this?)
Basically `get_point_intepolated()` would use the integer part to retrieve the exact pixel, then use the decimal part as a weight to blend with all height neighbors. I'm afraid of the performance impact since it needs to sample 9 points instead of 1, so I'm not sure if this should be in the heightmap code or in the mesh code. Let me know what you think.

**Comment by BryceBubbles** [comment]
Yeah I thought about adding interpolation too, but decided it wasn't worth it for now.

In general texture size should be the same or larger than mesh subdivide, otherwise you're just adding extra triangles for no reason right?

There's a couple of reasons why the examples look different. The mesh node has smoothing off, so you're seeing flat normals. And the shader is sampling from a filtered texture which adds some interpolation automatically.

That gives me an idea actually. Use get_image on the heightmap, and then use Godots image resize method with whatever interpolation setting looks best. Either size up to the subdivide value, or maybe resizing to the same size will work and be faster? Then pull the data back out with get_pixel.


That aside, the viewport setup I'm working on will allow 1-4k heightmaps and normalmaps to generate pretty much instantly. All calculations have to be moved into canvas shaders though.

I got a good chunk of this sorted today. The heightmap class will be replaced by a spatial graph input, with the viewports as children (it's better if they're always in the scene tree). Then only the input name is passed through nodes, in order to bypass duplication.

This means that branching to separate outputs won't work as expected for now, but eventually we can duplicate the viewports specifically when branching. Not sure on the specifics of that though. 

Lastly,  I'm currently unable to get the viewports to show in the scene tree, even when setting owner (but they still work as expected).
Output sometimes doesn't show too, even when the setting is checked in the graph. 
Is this a bug you're aware of?

There's also a bunch of gdnative errors when loading Godot. Let me know if you need details.

**Comment by HungryProton** [comment]
Okay, so worth it for now then!
I temporarily removed the gdnative part for now, you shouldn't have these errors anymore.

The output node that doesn't show though, no I've never had this issue before, I'll look into it.
